src/arch/mod.rs

```diff
@@ -517,3 +510,34 @@ impl From<usize> for Width {
         }
     }
 }
+
+// ———————————————————————— Helpers ————————————————————————— //
+
+/// Change mstatus.MPP and return the previous mstatus.MPP
+pub unsafe fn set_mpp(mode: Mode) -> Mode {
+    let value = mode.to_bits() << mstatus::MPP_OFFSET;
+    let prev_mstatus = Arch::read_csr(Csr::Mstatus);
+    Arch::write_csr(Csr::Mstatus, (prev_mstatus & !mstatus::MPP_FILTER) | value);
+    parse_mpp_return_mode(prev_mstatus)
+}
+
+/// Return the faulting instruction at the provided exception PC.
+///
+/// # Safety
+///
+/// The trap info must correspond to a valid trap info, no further checks are performed.
+pub unsafe fn get_raw_faulting_instr(trap_info: &TrapInfo) -> usize {
+    if trap_info.mcause == MCause::IllegalInstr as usize {
+        // First, try mtval and check if it contains an instruction
+        if trap_info.mtval != 0 {
+            return trap_info.mtval;
+        }
+    }
+
+    let instr_ptr = trap_info.mepc as *const u32;
+
+    // With compressed instruction extention ("C") instructions can be misaligned.
+    // TODO: add support for 16 bits instructions
+    let instr = ptr::read_unaligned(instr_ptr);
+    instr as usize
+}
```

Comment:
This will fault when run in user-space. We probably won't run into those kinds of issues until we verify the instruction fetch, so its fine for now, but lets keep that in mind if it shows up in the future.

Reply:
@CharlyCst Good to know. This commit was a simple refactoring and doesn't change the userspace function. Let me check if I find a way to formally verify the decoder